Add Linux support

tschuy opened the pull request:
Like you noted in #11, this essentially boils down to putting the ``app.dock`` into an if statement. (I don't think there's any Linux equivalent to the dock icon with a number on it, or the bouncing dock icon.)

I also rearranged the menu to be more cross-platformy, since there's no appname menu on most distributions of Linux. (Though there is in Gnome 3, the desktop manager I use, but I couldn't figure out how to add to it from Electron, and Atom doesn't use it either.)

There is a bug with desktop notifications crashing (https://github.com/atom/electron/issues/2294) that seems pretty awful. The issue isn't entirely clear which versions of what it presents itself on, so I added a note to the README about it. The notifications themselves work just fine, and as long as you don't click on them, the system won't crash.

I would be happy to make any changes you need before merging this. I could also test it on a non-Gnome environment.

sindresorhus commented on menu.js:
```diff
@@ -175,4 +105,84 @@ ${process.platform} ${process.arch} ${os.release()}`;
+		}
+	}
+);
+
```
Since the menues are so different on OS X and Linux I think it would be better to just create a separate file with a Linux menu instead of juggling this mess. The `Help` menu is common and can be a `menu-common.js` file and required into each.

tschuy replied:
That's reasonable. It started out not too bad, and then as I added more and more to it, it got crazier and crazier :stuck_out_tongue:

sindresorhus commented on readme.md:
```diff
@@ -5,10 +5,11 @@
 <br>
 [![](media/screenshot.png)](https://github.com/sindresorhus/caprine/releases/latest)
 
+*Requires OS X 10.8 or newer and Linux. Windows support planned.*
```
```
*Requires OS X 10.8+ or Linux. Windows support planned.*
```

sindresorhus commented on readme.md:
```diff
@@ -5,10 +5,11 @@
+## Installing on OS X
 
-*Requires OS X 10.8 or newer. Linux and Windows support planned.*
+*To run on Linux, clone the source and follow the ``dev`` instructions below.*
```
Why so manual? We can package for Linux using `electron-packager` as described in the Linux issue.

sindresorhus commented on readme.md:
```diff
@@ -34,6 +35,11 @@ Desktop notifications can be turned on in Preferences.
 
 <div align="center"><img src="media/screenshot-notification.png" width="358"></div>
 
+NOTE: There is a <a href="https://github.com/atom/electron/issues/2294">known bug</a>
```
Use Markdown link syntax

sindresorhus commented on readme.md:
```diff
@@ -34,6 +35,11 @@ Desktop notifications can be turned on in Preferences.
+with Electron's handling of desktop notifications on systems running Gnome 3
+that may cause Caprine to crash if notifications are clicked. Until this bug is
+resolved, do not click on notifications if they cause your the app to crash on
+your system.
```
Don't hard-wrap

sindresorhus commented on index.js:
```diff
@@ -13,6 +13,10 @@ require('crash-reporter').start();
 let mainWindow;
 
 function updateBadge(title) {
+	// don't try to update the dock icon if the system has no dock
```
I don't think this needs a comment. The code is self-explanatory.

sindresorhus commented:
Make sure the Linux menu is consistent with other Linux apps.

sindresorhus commented on menu.js:
```diff
@@ -175,4 +105,84 @@ ${process.platform} ${process.arch} ${os.release()}`;
+		type: 'separator'
+	},
+	{
+		label: 'Preferences...',
```
Re https://github.com/sindresorhus/caprine/pull/13#issuecomment-141874421

On Linux it's only `Preferences`, not `Preferences...`.

tschuy commented:
Unfortunately, ``About`` goes under help on Linux, so I can't actually factor out the help menu. :cry: 

From what I managed to get by messing around with ``electron-packager`` the best I'll be able to do is make an executable that can then be zipped up, it doesn't have a way to create a single runnable file. I'll make the script throw the build into a zip.

tschuy commented:
what would you prefer the behavior of ``npm run build`` be? it's constrained by running ``/bin/sh``, not ``/bin/bash``, so I can't do something fancy like ``if [[ "$OSTYPE" == "linux-gnu" ]]; then npm run build-linux; else npm run build-osx; fi``. I could write a separate ``build.sh`` script that compiles your current system, or I could just make it run OS X by default, or both?

sindresorhus commented:
`npm run build` should build for both platforms. `electron-packager` can handle this fine. See its docs.

tschuy commented:
When trying to build them together, I run into the following problems:

* ``sign`` doesn't exist on Linux, but ``electron-packager`` still tries to use the ``codesign`` executable, so it crashes
* it doesn't appear to want to put the executable into an archive, even when ``-asar`` is set
* for whatever reason, instead of setting the app icon at compile time, it's set in the source code as the ``icon`` value passed to ``BrowserWindow``. However, ``media/`` isn't included in the build, so there's no static file to reference.

I think two separate build commands might be necessary for that top reason. For the bottom two, I'm sure it's just me being silly and missing something in their documentation. (I still have no idea what ``asar`` is supposed to look like when it's successfully outputted.)

sindresorhus commented:
> sign doesn't exist on Linux, but electron-packager still tries to use the codesign executable, so it crashes

sindresorhus commented:
> sign doesn't exist on Linux, but electron-packager still tries to use the codesign executable, so it crashes

Ugh, `electron-packager` is so buggy. Wish we had something better...

I'm ok with just being able to build from OS X. I'm the only one that creates builds anyways. Having one build massively simplifies it, so I think it's worth the limitation.

> it doesn't appear to want to put the executable into an archive, even when -asar is set

Can you elaborate? It's the `zip` command that does this, not `electron-packager`.

> for whatever reason, instead of setting the app icon at compile time, it's set in the source code as the icon value passed to BrowserWindow. However, media/ isn't included in the build, so there's no static file to reference.

The media folder is ignored. I guess we can ignore everything in the media folder, except the icon.

tschuy commented:
> Can you elaborate?

from ``electron-packager``'s docs, it reads like ``asar`` should put everything into an archive: https://github.com/maxogden/electron-packager#usage

for now, I've just set the ``build-linux`` command to stick it in there itself.

What I could do is have ``build-linux`` and ``build-osx`` commands, and just have ``build`` be ``npm run build-osx && npm run build-linux``. This would still be one command, and it'd allow anyone that wants to run builds themselves (though you're right that people working on the project have no reason to build, only you) to just run the build they want. I don't know if this would break some caching mechanism that ``electron-packager`` has, though.

I tried removing ``media/`` from the build command for Linux and it's not included in the ``dist/Caprine-linux-x64`` folder, so I'm going to look more into it.

tschuy commented:
ok, I have a solution that more or less works, as far as I can tell. It should allow you to create a Linux zip and an OS X zip by running ``build``, and allow others to create zips for either system by running ``build-osx`` or ``build-linux``. The Linux build has a functioning icon, and I included a ``.desktop`` template in the README for anyone who wants to create a shortcut. Since this isn't using a proper install mechanism (which is to say, a deb or rpm), installation is a little clunky.

sindresorhus commented on menu.js:
```diff
@@ -10,169 +10,279 @@ function sendAction(action) {
-${app.getName()} ${app.getVersion()}
-${process.platform} ${process.arch} ${os.release()}`;
+		${app.getName()} ${app.getVersion()}
+		${process.platform} ${process.arch} ${os.release()}`;
```
Don't indent this. Template literals are whitespace sensitive.

sindresorhus commented on readme.md:
```diff
@@ -5,12 +5,11 @@
+*Requires OS X 10.8+ and Linux. Windows support planned.*
 
-## Install
+## Installing
```
Keep it `Install`

sindresorhus commented on menu.js:
```diff
@@ -10,169 +10,279 @@ function sendAction(action) {
+			]
+		}
+	];
+} else {
```
```
} else if (process.platform === 'linux') {
```

sindresorhus commented on menu.js:
```diff
@@ -10,169 +10,279 @@ function sendAction(action) {
+					}
+				},
+				{
+					label: `Quit ${appName}`,
```
AFAIK, linux apps don't have this.

sindresorhus commented on menu.js:
```diff
@@ -10,169 +10,279 @@ function sendAction(action) {
+					label: 'Select All',
+					accelerator: 'CmdOrCtrl+A',
+					role: 'selectall'
+				},
```
Linux apps doesn't have all of these either. Look at which one it does have.

sindresorhus commented on menu.js:
```diff
@@ -10,169 +10,279 @@ function sendAction(action) {
+					}
+				},
+				{
+					label: 'Report Issue',
```
These two submenu items shouldn't be duplicated. Put them in an external array and concat them into each of the menues.

sindresorhus commented on package.json:
```diff
@@ -17,7 +17,9 @@
     "start": "electron .",
-    "build": "electron-packager . $npm_package_productName --overwrite --out=dist --ignore='^/dist$' --ignore='^/media$' --prune --asar --platform=darwin --arch=x64 --icon=media/Icon.icns --app-bundle-id=com.sindresorhus.caprine --sign='Developer ID Application: Sindre Sorhus (YG56YK5RN5)' --app-version=$npm_package_version --version=$npm_package_electronVersion && cd dist/Caprine-darwin-x64 && z ...
+    "build": "npm run build-osx && npm run build-linux",
+    "build-osx": "electron-packager . $npm_package_productName --overwrite --out=dist --ignore='^/dist$' --ignore='^/media$' --prune --platform=darwin,linux --arch=x64 --icon=media/Icon.icns --app-bundle-id=com.sindresorhus.caprine --sign='Developer ID Application: Sindre Sorhus (YG56YK5RN5)' --app-version=$npm_package_version --version=$npm_package_electronVersion && cd dist/Caprine-darwin-x64 & ...
```
You need to remove `linux` here.

sindresorhus commented on menu.js:
```diff
@@ -10,169 +10,279 @@ function sendAction(action) {
-					sendAction('show-preferences');
+let tpl;
+
+if (process.platform === 'darwin') {
```
I'd rather have the two menues in two different variables and do the conditional at the bottom with just those variables.

sindresorhus commented on menu.js:
```diff
@@ -10,169 +10,279 @@ function sendAction(action) {
+					}
+				},
+				{
+					label: `About ${appName}`,
```
On Ubuntu it's just `About`.

Look at the existing Linux menues to make sure everything is consistent.

sindresorhus commented:
Is there any way to have the app installable on Linux with something like `Homebrew Cask` on OS X?

If https://github.com/loopline-systems/electron-builder supported Linux we could use that. Maybe worth opening a feature request for that?

battlemidget commented:
We could look at how atom builds their deb's/rpm's and mimic that. Would at least support the major distributions

tschuy commented:
I've addressed your feedback about the menus and fixed the accidental inclusion of linux in the build-osx command. 

For menu consistency, I'm not sure there's a really good standard compared to what there is on OS X. I tried to go off of Gedit. Most Gnome apps use the menu at the top of the screen (that I've not seen any other app use, and don't think there's support for in electron), but gedit has a more traditional File / Edit / Help etc.

You mentioned that the edit menu shouldn't have some of those items, but gedit has all of them. I've left them there for now; which ones were you thinking of?

About packaging -- I don't think there's a good way to distribute this unless we make it build an RPM and a DEB. I'll make a feature request on https://github.com/loopline-systems/electron-builder, and revisit this later if they add support.

tschuy commented:
Also, I'd be more than happy to squash these commits if you want so that ``address feedback`` isn't in the git log.

tschuy commented:
issue for adding linux support to the builder: https://github.com/loopline-systems/electron-builder/issues/7

battlemidget commented:
> You mentioned that the edit menu shouldn't have some of those items, but gedit has all of them. I've left them there for now; which ones were you thinking of?

I think it just depends on the application, things like gedit, kate, mousepad, etc make use of those edit menu items where as others don't. They probably are useful for any type of text editing but I'm not sure this app needs menu items for it. Take for example Telegram on the desktop for Linux, it doesn't contain any of those menu items (File, Edit) etc.

sindresorhus commented:
@tschuy Landed! Really appriciate you working on this @tschuy :)
